eng/DownloadSharedSource.ps1

```diff
@@ -2,29 +2,89 @@
 $ErrorActionPreference = 'Stop'
 $ProgressPreference = 'SilentlyContinue'
 
-function DownloadAll([string[]]$files, [string]$baseUrl, [string]$downloadPath)
-{
-    foreach ($file in $files)
-    {
-        Write-Host "Downloading" $file
-        $text = (Invoke-WebRequest -Uri "$baseUrl/$file").Content
-        $text.Trim() | Out-File (Join-Path $downloadPath $file)
+$sparseCheckoutFile = ".git/info/sparse-checkout"
+$gitRemoteUrl = "https://github.com/Azure/azure-sdk-for-net.git"
+$gitRemoteBranchName = "main"
+$azCoreSharedPath = "sdk/core/Azure.Core/src/Shared/"
+$armCoreSharedPath = "sdk/resourcemanager/Azure.ResourceManager/src/Shared/"
+
+function InitializeSparseGitClone([string]$repo) {
+    git clone --no-checkout --filter=tree:0 $repo .
+    if ($LASTEXITCODE) { exit $LASTEXITCODE }
+    git sparse-checkout init
+    if ($LASTEXITCODE) { exit $LASTEXITCODE }
+    Remove-Item $sparseCheckoutFile -Force
+}
+
+function AddSparseCheckoutPath([string]$subDirectory) {
+    if (!(Test-Path $sparseCheckoutFile) -or !((Get-Content $sparseCheckoutFile).Contains($subDirectory))) {
+        Write-Output $subDirectory >> $sparseCheckoutFile
+    }
+}
+
+function GetSDKCloneDir() {
+    $root = git rev-parse --show-toplevel
+
+    $sparseSpecCloneDir = "$root/sdk"
```

Comment:
Can we use a place out of this repo?

Reply:
Yes lets put this somewhere not inside the existing repo.

Reply:
This will be run in 2 pipelines where the current repo is one of several in the agent work folder and possibly on dev machines.  I wouldn't write outside of the repo on dev machines unless we take the location as a parameter.

Reply:
> I'd recommend we put it in a gitignored folder within the repo. I think under `/artifacts` is appropriate.

This sound reasonable.
@pshao25 @weshaggard I do not want to put it outside this repo because I do not want to introduce the possibility that the script in this repo accidentally changes something outside this repo because those might be irreversible.